### PR TITLE
Introduce mimetype DB update occ command

### LINKED ISSUE #17625: mimetype mapping occ command needed
Now that we have mimetypemapping.dist.json in /conf we should add a command to occ to update all the mimetypes in the database.

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
This occ command allows mimetypes in the database to be updated from changed mimetypes in config/mimetypemapping.json. By default, it adds any new mimetypes it finds and updates the filecache based on those, but with an option it can scan all existing mimetypes to update mimetypes on files.

For example, you add a new mapping of '.foo' to 'application/foobar', but your existing '.foo' files still have the 'application/octet-stream' mimetype since they were uploaded before. Run `./occ maintenance:mimetype:update-db` to update this. Or, if the 'application/foobar' mimetype already existed, `./occ maintenance:mimetype:update-db --repair-filecache` (will be a lot slower for big filecaches!).

Please review @rullzer @icewind1991 @MorrisJobke @PVince81 @icewind1991 @DeepDiver1975 

cc @carlaschroder for documentation around the new occ command, and rename of the maintenance:mimetypesjs command to maintenance:mimetype:update-js

Fixes #17625

### CODE REVIEW

**Comment by karlitschek** [comment]
good idea :+1:

**Comment by icewind1991** [comment]
:+1: code looks good, saw it work

**Comment by rullzer** [comment]
Looking good!
And works!
:+1: 

But we really should have tests for this as well (I know I did not write them either...)
Also remember to also update the existing docs since the command is changed.

File: core/register_command.php [hunk 1]
```diff
@@ -79,10 +79,11 @@
 	);
 	$application->add(new OC\Core\Command\Encryption\ShowKeyStorageRoot($util));
 
-	$application->add(new OC\Core\Command\Maintenance\MimeTypesJS());
 	$application->add(new OC\Core\Command\Maintenance\Mode(\OC::$server->getConfig()));
 	$application->add(new OC\Core\Command\Maintenance\Repair(new \OC\Repair(\OC\Repair::getRepairSteps()), \OC::$server->getConfig()));
 	$application->add(new OC\Core\Command\Maintenance\SingleUser(\OC::$server->getConfig()));
+	$application->add(new OC\Core\Command\Maintenance\Mimetype\UpdateJS(\OC::$server->getMimeTypeDetector()));
```
Comment (nickvergessen) [inline core/register_command.php]: Alphabetic order :speak_no_evil:

Comment (Xenopathic) [inline core/register_command.php]: You mean like the rest of the file ~~is~~? :laughing:

**Comment by Xenopathic** [comment]
Well, doing it properly turned out to be a lot more work than expected, but at least some code smell has been removed. Everything is unit tested too

**Comment by icewind1991** [comment]
Looks good :+1:

**Comment by rullzer** [comment]
Looks good and everything seems to work. :+1: 
Thanks for the tests :)

**Comment by oparoz** [comment]
Awesome! :)

**Comment by carlaschroder** [comment]
But I like code smell. It smells like victory.

**Comment by nickvergessen** [comment]
Regression: master unit tests are broken:
bildschirmfoto vom 2015-09-11 13 23 36

https://github.com/owncloud/core/issues/18975

**Pull request merged**